corgi/tasks/managed_services.py

```diff
@@ -0,0 +1,83 @@
+from celery_singleton import Singleton
+from django.db import transaction
+from django.utils import timezone
+
+from config.celery import app
+from corgi.collectors.app_interface import AppInterface
+from corgi.collectors.syft import Syft
+from corgi.core.models import Component, ComponentNode, ProductStream, SoftwareBuild
+from corgi.tasks.common import RETRY_KWARGS, RETRYABLE_ERRORS
+from corgi.tasks.sca import save_component
+
+
+@app.task(base=Singleton, autoretry_for=RETRYABLE_ERRORS, retry_kwargs=RETRY_KWARGS)
+def refresh_service_manifests():
+    services = ProductStream.objects.filter(meta_attr__managed_service_components__isnull=False)
+    service_metadata = AppInterface.fetch_service_metadata(services)
+
+    for service, components in service_metadata.items():
+        manifest_service(service.pk, components)
+        manifest_service.delay(service.pk, components)
+
+
+@app.task(base=Singleton, autoretry_for=RETRYABLE_ERRORS, retry_kwargs=RETRY_KWARGS)
+@transaction.atomic
+def manifest_service(product_stream_id: str, service_components: list):
+    service = ProductStream.objects.get(pk=product_stream_id)
+
+    touched_build_ids = []
+    now = timezone.now()
+    for service_component in service_components:
+        analyzed_components = []
+        quay_repo = service_component.get("quay_repo_name")
+        if quay_repo:
+            quay_repo_full = f"quay.io/{quay_repo}"
+            component_data, scan_source = Syft.scan_repo_image(target_image=quay_repo_full)
+            for scanned_component in component_data:
+                scanned_component["meta"][
+                    "quay_image"
+                ] = f"{quay_repo_full}@{scan_source['target']['imageID']}"
+            analyzed_components.extend(component_data)
+
+        git_repo = service_component.get("git_repo_url")
+        if git_repo:
+            component_data, source_ref = Syft.scan_git_repo(target_url=git_repo)
+            for scanned_component in component_data:
+                scanned_component["meta"]["git_repo"] = f"{git_repo}@{source_ref}"
+            analyzed_components.extend(component_data)
+
+        if not analyzed_components:
+            continue
+
+        faux_version = now.strftime("%Y%m%d%H%M%S")
```

Comment:
What we last ingested is as good as it gets since we only check periodically (thus the scheduled task). We don't have a great way of discovering newly deployed versions so daily checks is good enough (this is how the current manifesting through mangen works as well).

Reply:
I also originally thought about wiping out the previous runs data prior to loading a new set, but figured maybe we do want to keep a history of changes.

Reply:
Ok, let's make sure that when we call 'get_latest_components' for these product streams we get the latest root_components.

Reply:
I actually did add the deletion code for past builds because otherwise we'd have an ever-growing mess of components linked to many builds, but with get_or_create the past builds would simply be getting orphaned unless we made their versions unique every time somehow (even though in reality they may not have changed).

I guess this will solve itself once we transition to a properly tracked history against specific builds instead of this periodic refresh.